buffer/src/main/java/io/netty/buffer/Unpooled.java

```diff
@@ -262,38 +262,86 @@ public static ByteBuf wrappedBuffer(ByteBuffer... buffers) {
         return wrappedBuffer(buffers.length, buffers);
     }
 
+    // used to avoid temporary array allocation/copy when wrapping byte[][] and ByteBuffer[]
+    abstract static class WrappingIterable<T> implements Iterable<ByteBuf> {
+        private final T[] array;
+        private final int start;
+
+        protected WrappingIterable(T[] array) {
+            this.array = array;
+
+            for (int i = 0;; i++) {
+                if (i >= array.length || array[i] == null) {
+                    start = array.length;
+                    return;
+                }
+                if (!isEmpty(array[i])) {
+                    start = i;
+                    return;
+                }
+            }
+        }
+
+        int maxSize() {
+            return array.length - start;
+        }
+
+        @Override
+        public Iterator<ByteBuf> iterator() {
+            return new Iterator<ByteBuf>() {
+                int i = start;
+
+                @Override
+                public boolean hasNext() {
+                    for (; i < array.length && array[i] != null; i++) {
```

Comment:
You should assign `array[I]` to something and re-use it, as at the moment you access each element two times.